runtime/src/accounts_db.rs

```diff
@@ -713,26 +757,56 @@ impl AccountsDB {
         // Calculate store counts as if everything was purged
         // Then purge if we can
         let mut store_counts: HashMap<AppendVecId, (usize, HashSet<Pubkey>)> = HashMap::new();
-        let storage = self.storage.read().unwrap();
-        for (key, (account_infos, _ref_count)) in &purges {
-            for (slot, account_info) in account_infos {
-                let slot_storage = storage.0.get(&slot).unwrap();
-                let store = slot_storage.get(&account_info.store_id).unwrap();
+        for (key, (account_infos, ref_count)) in purges.iter_mut() {
+            if purged_account_slots.contains_key(&key) {
+                *ref_count = self
+                    .accounts_index
+                    .read()
+                    .unwrap()
+                    .ref_count_from_storage(&key);
+            }
+            account_infos.retain(|(slot, account_info)| {
+                let was_slot_purged = purged_account_slots
+                    .get(&key)
+                    .map(|slots_removed| slots_removed.contains(slot))
+                    .unwrap_or(false);
+                if was_slot_purged {
+                    // No need to look up the slot storage below if the entire
+                    // slot was purged
+                    return false;
+                }
+                // Check if this update in `slot` to the account with `key` was reclaimed earlier by
+                // `clean_old_rooted_accounts()`
+                let was_reclaimed = removed_accounts
+                    .get(&account_info.store_id)
+                    .map(|store_removed| store_removed.contains(&account_info.offset))
+                    .unwrap_or(false);
+                if was_reclaimed {
+                    return false;
+                }
                 if let Some(store_count) = store_counts.get_mut(&account_info.store_id) {
                     store_count.0 -= 1;
                     store_count.1.insert(*key);
                 } else {
                     let mut key_set = HashSet::new();
                     key_set.insert(*key);
-                    store_counts.insert(
-                        account_info.store_id,
-                        (store.count_and_status.read().unwrap().0 - 1, key_set),
+                    let count = self
+                        .storage
+                        .read()
```

Comment:
@carllin Opps, Sorry to spot this at later stage of this pr. But, I think there is `panic!` possibility here, because of the shrinker thread, which is running in parallel.

By moving this `self.storage.read()` lock inside here from the outer loop prologue, now shrinker thread are free to update account storages while this is running. `unwrap()` at L798 here bite us.

This is because shrinker  effectively replaces old AccountStorageEntry with new one while removing the old AccountStorageEntry (we might be accessing this from here). `remove_dead_accounts()` never cleans up any empty AccountStorageEntries unless all of them for given slot is empty. But the `do_shrink_slot` do its work diligently by removing any empty StorageEntry manually after `handle_reclaims`, knowing they can almost certainly can remove empty one of them at the very least.

As a matter of fact, we're not affected from other store codepath, however. Because we're mangling rooted slots. other store codepath only touches non-rooted new slots. The only two which touches the rooted slots are this cleaner and shrinker.

Reply:
@ryoqun, shrink and clean run serially in AccountsBackgroundService, no?

Reply:
Yeah, you're right; Even more, the above comment is wrong: we're actually serializing the executions: https://github.com/solana-labs/solana/pull/12606/files?diff=split&w=1#diff-2099c5256db4eb5975c8834af38f6456R703

nis: So, I think this lock can go away now.